enterprise/backend/src/metabase_enterprise/caching/strategies.clj

```diff
@@ -0,0 +1,218 @@
+(ns metabase-enterprise.caching.strategies
+  (:require
+   [clojurewerkz.quartzite.jobs :as jobs]
+   [clojurewerkz.quartzite.schedule.cron :as cron]
+   [clojurewerkz.quartzite.triggers :as triggers]
+   [java-time.api :as t]
+   [malli.core :as mc]
+   [medley.core :as m]
+   [metabase.public-settings :as public-settings]
+   [metabase.public-settings.premium-features
+    :refer [defenterprise defenterprise-schema]]
+   [metabase.query-processor :as qp]
+   [metabase.query-processor.middleware.cache-backend.db :as backend.db]
+   [metabase.task :as task]
+   [metabase.util.cron :as u.cron]
+   [metabase.util.log :as log]
+   [metabase.util.malli.schema :as ms]
+   [toucan2.core :as t2])
+  (:import
+   (org.quartz.spi MutableTrigger)))
+
+(set! *warn-on-reflection* true)
+
+(def CacheStrategy ;; not used in API since there are some additional fields
+  "Schema for a caching strategy"
+  (mc/schema
+   [:and
+    [:map
+     [:type [:enum :nocache :ttl :duration :schedule :query]]]
+    [:multi {:dispatch :type}
+     [:nocache  [:map
+                 [:type keyword?]]]
+     [:ttl      [:map
+                 [:type keyword?]
+                 [:multiplier ms/PositiveInt]
+                 [:min_duration ms/PositiveInt]]]
+     [:duration [:map
+                 [:type keyword?]
+                 [:duration ms/PositiveInt]
+                 [:unit [:enum "hours" "minutes" "seconds" "days"]]]]
+     [:schedule [:map
+                 [:type keyword?]
+                 [:updated_at some?]
+                 [:schedule u.cron/CronScheduleString]]]
+     [:query    [:map
+                 [:type keyword?]
+                 [:updated_at some?]
+                 [:payload [:maybe any?]]
+                 [:field_id int?]
+                 [:aggregation [:enum "max" "count"]]
+                 [:schedule string?]]]]]))
+
+(defenterprise refreshable-states
+  "States of `persisted_info` records which can be refreshed."
+  :feature :cache-granular-controls
+  []
+  #{"creating" "persisted" "error"})
+
+(defenterprise prunable-states
+  "States of `persisted_info` records which can be pruned."
+  :feature :cache-granular-controls
+  []
+  #{"deletable" "off"})
+
+(defn granular-ttl
+  "Returns the granular cache ttl (in seconds) for a card. On EE, this first checking whether there is a stored value
+   for the card, dashboard, or database (in that order of decreasing preference). Returns nil on OSS."
+  [card dashboard database]
+  (let [ttls              [(:cache_ttl card) (:cache_ttl dashboard) (:cache_ttl database)]
```

Comment:
nit: we can optimize this a big since we only care about first non-nil ttl. I think we can eliminate unnecessary call to get `dashboard`, `database` if `card` already have a ttl.

Reply:
nice; even though I'd love to migrate and remove this code, that'll remove a few queries maybe. Funny part is that in old version those queries were done even in OSS version.